File: generated_code/GeminiPro/Python/092_SplayTree.py
class SplayTree:
    """
    A self-contained Python class implementing a Splay Tree.

    This class provides a dictionary-like set for integers, supporting
    insert, delete, and search operations. The key feature is the splaying
    operation, which moves accessed nodes to the root to optimize
    future accesses.
    """

    class _Node:
        """A private inner class for the nodes of the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splay operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig step
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig step
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig step
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            else:  # Zig-Zag step
                if node == parent.right and parent == grandparent.left:
                    self._left_rotate(parent)
                    self._right_rotate(grandparent)
                else:
                    self._right_rotate(parent)
                    self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        Performs a splay operation on the found node or the last accessed
        node if the key is not found.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        last_visited = None
        current = self.root
        while current:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:  # Key found
                self._splay(current)
                return True
        
        # Key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
        
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the node is splayed to the root.
        If the key does not exist, it is inserted and then splayed to the root.

        Args:
            key: The integer key to insert.
        """
        # Case 1: Tree is empty
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the closest node to the key. If key exists, it becomes the root.
        if self.search(key):
            return  # Key already exists, search has splayed it.

        # After a failed search, the root is the closest leaf/node.
        # We insert the new key as the new root and restructure.
        new_node = self._Node(key)
        old_root = self.root

        if key < old_root.key:
            new_node.right = old_root
            new_node.left = old_root.left
            if old_root.left:
                old_root.left.parent = new_node
            old_root.left = None
        else:  # key > old_root.key
            new_node.left = old_root
            new_node.right = old_root.right
            if old_root.right:
                old_root.right.parent = new_node
            old_root.right = None
        
        old_root.parent = new_node
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is found, it is deleted. The tree is then restructured
        by splaying the largest element in the left subtree. If the key is
        not found, the tree structure is modified by the preceding search
        and splay operation.

        Args:
            key: The integer key to delete.
        """
        # Search for the key. This will splay it to the root if it exists.
        if not self.search(key):
            # Key is not in the tree, nothing to delete.
            # search() has already splayed the closest node.
            return

        # At this point, the node to delete is the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left child, the right subtree becomes the new tree.
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect the left subtree from the old root.
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max_node. To do this within the left subtree,
            # we temporarily set it as the main root.
            self.root = left_subtree
            self._splay(max_node)
            
            # The splayed max_node is now the root. It has no right child.
            # Attach the original right subtree to it.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/092_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * Splay trees are self-balancing binary search trees with the property that recently
 * accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * It is a private static inner class to encapsulate the node structure.
     */
    private static class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Private Helper Methods for Rotations and Splaying ---

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * This is a standard BST rotation.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rotateRight(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * This is a standard BST rotation.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rotateLeft(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splay operation. It brings the node with the given key to the root of the tree.
     * If the key is not present, the last accessed node in the search path is brought to the root.
     * This method uses a recursive, bottom-up approach.
     *
     * @param currentRoot The root of the current subtree being considered.
     * @param key The key to be splayed to the top.
     * @return The new root of the tree after the splay operation.
     */
    private Node splay(Node currentRoot, int key) {
        // Base case: root is null or key is present at root
        if (currentRoot == null || currentRoot.key == key) {
            return currentRoot;
        }

        // Key lies in the left subtree
        if (key < currentRoot.key) {
            // Key is not in the tree, we are done
            if (currentRoot.left == null) {
                return currentRoot;
            }

            // Zig-Zig (Left-Left)
            if (key < currentRoot.left.key) {
                // Recursively bring the key as root of the left-left grandchild
                currentRoot.left.left = splay(currentRoot.left.left, key);
                // Perform the first right rotation for the current root
                currentRoot = rotateRight(currentRoot);
            }
            // Zig-Zag (Left-Right)
            else if (key > currentRoot.left.key) {
                // Recursively bring the key as root of the left-right grandchild
                currentRoot.left.right = splay(currentRoot.left.right, key);
                // Perform the first rotation (left) for the child
                if (currentRoot.left.right != null) {
                    currentRoot.left = rotateLeft(currentRoot.left);
                }
            }

            // Perform the second rotation for the current root (if child exists)
            return (currentRoot.left == null) ? currentRoot : rotateRight(currentRoot);
        }
        // Key lies in the right subtree
        else {
            // Key is not in the tree, we are done
            if (currentRoot.right == null) {
                return currentRoot;
            }

            // Zig-Zag (Right-Left)
            if (key < currentRoot.right.key) {
                // Recursively bring the key as root of the right-left grandchild
                currentRoot.right.left = splay(currentRoot.right.left, key);
                // Perform the first rotation (right) for the child
                if (currentRoot.right.left != null) {
                    currentRoot.right = rotateRight(currentRoot.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > currentRoot.right.key) {
                // Recursively bring the key as root of the right-right grandchild
                currentRoot.right.right = splay(currentRoot.right.right, key);
                // Perform the first left rotation for the current root
                currentRoot = rotateLeft(currentRoot);
            }

            // Perform the second rotation for the current root (if child exists)
            return (currentRoot.right == null) ? currentRoot : rotateLeft(currentRoot);
        }
    }

    // --- Public API Methods ---

    /**
     * Inserts a key into the Splay Tree. If the key already exists, the node
     * containing the key is splayed to the root, but no new node is added.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // Case 1: The tree is empty
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. The closest node becomes the new root.
        root = splay(root, key);

        // Case 2: The key is already in the tree. Do nothing.
        if (root.key == key) {
            return;
        }

        // Case 3: The key is not in the tree. Insert a new node.
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Searches for a key in the tree. The accessed node (or the last node on the
     * search path if not found) is splayed to become the new root.
     *
     * @param key The key to search for.
     * @return The key as an Integer if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Deletes a key from the Splay Tree. If the key is found, it is first
     * splayed to the root and then removed.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key. If the key exists, it becomes the root.
        root = splay(root, key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The key to be deleted is now at the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // No left child, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // A search for the (just deleted) key will achieve this, as it's larger
            // than any element in the left subtree.
            Node newRoot = splay(leftSubtree, key);
            // The new root of the left subtree has no right child.
            // Attach the original right subtree there.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}